lib/Doctrine/ORM/Query/SqlWalker.php

```diff
@@ -854,9 +856,40 @@ public function walkIdentificationVariableDeclaration($identificationVariableDec
      */
     public function walkIndexBy($indexBy)
     {
-        $pathExpression = $indexBy->simpleStateFieldPathExpression;
+        $pathExpression = $indexBy->singleValuedPathExpression;
         $alias          = $pathExpression->identificationVariable;
-        $field          = $pathExpression->field;
+
+        /** @var Query\AST\PathExpression $pathExpression */
+        switch ($pathExpression->type) {
+            case AST\PathExpression::TYPE_STATE_FIELD:
+                $field = $pathExpression->field;
+                break;
+
+            case AST\PathExpression::TYPE_SINGLE_VALUED_ASSOCIATION:
+                // Just use the foreign key, i.e. u.group_id
+                $fieldName = $pathExpression->field;
+                $class     = $this->queryComponents[$alias]['metadata'];
+
+                if (isset($class->associationMappings[$fieldName]['inherited'])) {
+                    $class = $this->em->getClassMetadata($class->associationMappings[$fieldName]['inherited']);
+                }
+
+                $association = $class->associationMappings[$fieldName];
+
+                if (! $association['isOwningSide']) {
+                    throw QueryException::associationPathInverseSideNotSupported($pathExpression);
+                }
+
+                if (count($association['sourceToTargetKeyColumns']) > 1) {
+                    throw QueryException::associationPathCompositeKeyNotSupported();
+                }
+
+                $field = reset($association['targetToSourceKeyColumns']);
+                break;
+
+            default:
+                throw QueryException::invalidPathExpression($pathExpression);
+        }
```

Comment:
This seems like something that could be extracted to a private method and possibly use `if` and early exits (allowing for us to have a more strict comparison).

Reply:
Also, is this done elsewhere?

Reply:
This is largely a copy of the switch from walkPathExpression(), though not identical. The walkPathExpression() one is building a sql string wheras this one is just determining the appropriate field.
Should I try combining them both into a private method?